Tools/unicode/dawg.py

```diff
@@ -0,0 +1,486 @@
+                    s.append(c)
+                node.linear_edges.append((''.join(s), child))
+
+    def _topological_order(self):
+        # compute reachable linear nodes, and the set of incoming edges for each node
+        order = []
+        stack = [self.root]
+        seen = set()
+        while stack:
+            # depth first traversal
+            node = stack.pop()
+            if node.id in seen:
+                continue
+            seen.add(node.id)
+            order.append(node)
+            for label, child in node.linear_edges:
+                stack.append(child)
+
+        # do a (slightly bad) topological sort
+        incoming = defaultdict(set)
+        for node in order:
+            for label, child in node.linear_edges:
+                incoming[child].add((label, node))
+        no_incoming = [order[0]]
+        topoorder = []
+        positions = {}
+        while no_incoming:
+            node = no_incoming.pop()
+            topoorder.append(node)
+            positions[node] = len(topoorder)
+            # use "reversed" to make sure that the linear_edges get reorderd
+            # from their alphabetical order as little as necessary (no_incoming
+            # is LIFO)
+            for label, child in reversed(node.linear_edges):
+                incoming[child].discard((label, node))
+                if not incoming[child]:
+                    no_incoming.append(child)
+                    del incoming[child]
+        # check result
+        assert set(topoorder) == set(order)
+        assert len(set(topoorder)) == len(topoorder)
+
+        for node in order:
+            node.linear_edges.sort(key=lambda element: positions[element[1]])
+
+        for node in order:
+            for label, child in node.linear_edges:
+                assert positions[child] > positions[node]
+        # number the nodes. afterwards every input string in the set has a
+        # unique number in the 0 <= number < len(data). We then put the data in
+        # self.data into a linear list using these numbers as indexes.
+        topoorder[0].num_reachable_linear
+        linear_data = [None] * len(self.data)
+        inverse = {} # maps value back to index
+        for word, value in self.data.items():
+            index = self._lookup(word)
+            linear_data[index] = value
+            inverse[value] = index
+
+        return topoorder, linear_data, inverse
+
+    def compute_packed(self, order):
+        # assign offsets to every node
+        for i, node in enumerate(order):
+            # we don't know position of the edge yet, just use something big as
+            # the starting position. we'll have to do further iterations anyway,
+            # but the size is at least a lower limit then
+            node.packed_offset = 2 ** 30 + i * 2 ** 10
+
+        # due to the varint encoding of edge targets we need to run this to
+        # fixpoint
+        last_result = None
+        while 1:
```

Comment:
This is the biggest issue of your comments. it seems the delta-of-delta encoding of the distances can really lead to a bistable fixpoint. I fear this will require some careful fixing (the problem didn't occur for practical DBs, but that's not saying anything of course – would be very annoying if for unicode db 27 the script suddenly starts hanging :wink: ).

Reply:
Ok, I changed the fixpoint code to be much more structured and careful, fixing the bug. It's now no longer a correctness issue, instead every iteration shrinks the output until we can't shrink more, but we could stop at any earlier point and get a correct (but unnecessarily big) result.